### PR TITLE
Implement timezone interface (redux)

### PR DESCRIPTION
**Pull request by icambron**
This is a rebased version of #3134. I haven't done anything other than remove conflicts and fix tests.

See https://github.com/moment/moment-rfcs/pull/1

### CODE REVIEW

File: src/lib/create/from-array.js [hunk 1]
```diff
@@ -1,19 +1,16 @@
...
 
-function currentDateArray(config) {
+function currentDateArray(zone) {
     // hooks is actually the exported moment object
-    var nowValue = new Date(hooks.now());
-    if (config._useUTC) {
-        return [nowValue.getUTCFullYear(), nowValue.getUTCMonth(), nowValue.getUTCDate()];
-    }
-    return [nowValue.getFullYear(), nowValue.getMonth(), nowValue.getDate()];
+    var nowValue = hooks.withTimeZone(zone)()._d;
```
Comment (icambron) [inline src/lib/create/from-array.js]: This is the only real change I made to Tim's code. You can see his [here](https://github.com/moment/moment/pull/3134/files#diff-9ef3d17fb73729933240837060eeb3ecR12). I was lucky to have run the tests while my local date was different than UTC's or I'd have never seen it.

File: src/lib/create/from-anything.js [hunk 2]
```diff
@@ -46,12 +46,13 @@ export function prepareConfig (config) {
     if (isMoment(input)) {
         return new Moment(checkOverflow(input));
     } else if (isDate(input)) {
-        config._d = input;
+        config._d = new Date(input.valueOf());
+        config._offset = 0;
```
Comment (ichernev) [inline src/lib/create/from-anything.js]: Probably this should be a default, so we don't have to set it everywhere.

File: src/lib/create/utc.js [hunk 3]
```diff
@@ -1,5 +1,8 @@
-import { createLocalOrUTC } from './from-anything';
+import FixedOffsetTimeZone from '../timezone/fixed-offset';
+import { createWithTimeZone } from './from-anything';
 
-export function createUTC (input, format, locale, strict) {
-    return createLocalOrUTC(input, format, locale, strict, true).utc();
+var create = createWithTimeZone(new FixedOffsetTimeZone(0));
+
+export function createUTC () {
+    return create.apply(null, arguments).utcOffset(0);
```
Comment (icambron) [inline src/lib/create/utc.js]: I'd have to trace back through the details, but it seems like `utcOffset(0)` is actually a noop here? The line just above defines `create` to use an offset of 0.

File: src/lib/create/from-anything.js (hunk 2 above)
Comment (icambron) [inline src/lib/create/from-anything.js]: Yeah, I agree

**Comment by icambron** [comment]
I think I understood all of Tim's changes when I did this rebase, but I'd have to go back through to remember how everything works.

**Comment by ichernev** [comment]
I'm closing this, all of it is going in 3.0, with the immutability code and a few tweaks (namely API will be UTC -> offset in local time, not LocalTime -> offset in local (because its ambiguous)).

**Pull request closed without merging**